cookbook/controller/argument_value_resolver.rst

```diff
@@ -0,0 +1,157 @@
+
+
+Functionality Shipped With The HttpKernel
+-----------------------------------------
+
+Symfony ships with four value resolvers in the HttpKernel:
+ - The :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\ArgumentFromAttributeResolver`
+   attempts to find a request attribute that matches the name of the argument.
+
+ - The :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\RequestValueResolver`
+   injects the current ``Request`` if type-hinted with ``Request``, or a sub-class thereof.
+
+ - The :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\DefaultValueResolver`
+   will set the default value of the argument if present and the argument is optional.
+
+ - The :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\VariadicValueResolver`
+   verifies in the request if your data is an array and will add all of them to the argument list.
+   When the action is called, the last (variadic) argument will contain all the values of this array.
+
+.. note::
+    In older versions of Symfony this logic was all resolved within the ``ControllerResolver``. The
+    old functionality is moved to the ``LegacyArgumentResolver``, which contains the previously
+    used resolving logic.
+
+Adding a New Value Resolver
+---------------------------
+
+Adding a new value resolver requires one class and one service defintion. In our next example, we
+will be creating a shortcut to inject the ``User`` object from our security. Given we write the following
+action::
+
+    namespace AppBundle\Controller;
+
+    class UserController
+    {
+        public function indexAction(User $user)
+        {
+            return new Response('<html><body>Hello '.$user->getUsername().'!</body></html>');
+        }
+    }
+
+Somehow we will have to get the ``User`` object and inject it into our action. This can be done
+by implementing the :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolverInterface`.
+This interface specifies that we have to implement two methods::
+
+    interface ArgumentValueResolverInterface
+    {
+        public function supports(Request $request, ArgumentMetadata $argument);
+        public function resolve(Request $request, ArgumentMetadata $argument);
+    }
+
+ - The ``supports()`` method is used to check whether the resolver supports the given argument. It will
+   only continue if it returns ``true``.
+
+ - The ``resolve()`` method will be used to resolve the actual value just acknowledged by
+   ``supports()``. Once a value is resolved you can ``yield`` the value to the ``ArgumentResolver``.
+
+ - The ``Request`` object is the current ``Request`` which would also be injected into your
+   action in the forementioned functionality.
+
+ - The :class:``Symfony\\Component\\HttpKernel\\ControllerMetadata\\ArgumentMetadata`` represents
+   information retrieved from the method signature for the current argument it's trying to resolve.
```

Comment:
Lists shouldn't be indented with one space in reSt, it'll end up as a blockquote. Also, please use `*` to be consistent with the other lists in the documentation.